linux-kernel-rpi64-lts: new, 6.6.62

Dustymind opened the pull request:
<!-- For description on topic creation and maintenance, please refer to [this Wiki article](https://wiki.aosc.io/developer/packaging/topic-based-maintenance-guideline/). -->

Topic Description
-----------------

- linux-kernel-rpi64-lts: new, 6.6.62
- linux+kernel+rpi64+lts: new, 6.6.62
- rpi-firmware-boot: new, 1.20241126

Package(s) Affected
-------------------

- linux-kernel-rpi64-lts: 6.6.62
- linux+kernel+rpi64+lts: 6.6.62
- rpi-firmware-boot: 1.20241126

Security Update?
----------------

<!-- If this topic contains security update(s), please uncomment "Yes,"
     mark with the `security` and `priority` labels, and make sure to mark your commits to relevant issue numbers.

     Please see GitHub's documentation on "Linking a pull request to an issue":

     https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue -->

<!-- Yes - Issue Number: ISSUENUMBER -->
No

<!-- Please uncomment the "Build Order" section if your topic affects more than one package. -->

Build Order
-----------

<!-- Please describe in what order maintainers should build this pull request. You can use the following template, use space to separate packages. -->


```
#buildit rpi-firmware-boot linux-kernel-rpi64-lts linux+kernel+rpi64+lts
```

Test Build(s) Done
------------------

**Primary Architectures**

<!-- Please remove any architecture to which this topic does not apply. -->

- [x] AArch64 `arm64`

<!-- Maintainers should review file changes and make sure that the change(s) made complies with our [Package Styling Manual](https://wiki.aosc.io/developer/packaging/package-styling-manual/) -->

<!-- Maintainers and users may now test the packages in this topic and, once user/maintainer feedback indicates that the update(s) work as expected and find its quality satisfactory,
     another maintainer may now review this pull request and mark it as Approved. After which, the maintainer will build affected package(s) and upload them to the `stable` repository. -->

Dustymind committed:
rpi-firmware-boot: new, 1.20241126

Dustymind committed:
linux-kernel-rpi64-lts: new, 6.6.62

Dustymind committed:
linux-kernel-rpi64-lts: update config

    - Enable CONFIG_FW_LOADER_COMPRESS, CONFIG_FW_LOADER_COMPRESS_XZ, CONFIG_FW_LOADER_COMPRESS_ZSTD
    - Compile XFS filesystem support into kernel
    - Enable CONFIG_XFS_ONLINE_SCRUB and CONFIG_XFS_ONLINE_SCRUB_STATS

Dustymind committed:
linux-kernel-rpi64-lts: add missing config
    - add '# CONFIG_MODULE_UNLOAD_TAINT_TRACKING is not set'
    - add '# CONFIG_XFS_ONLINE_SCRUB_STATS is not set'
    - Enable CONFIG_XFS_ONLINE_REPAIR

Dustymind committed:
rpi-firmware-boot: add some config from upstream

Cyanoxygen commented on runtime-kernel/linux-kernel-rpi64-lts/autobuild/postinst.in:
```diff
@@ -0,0 +1,33 @@
+    # failed, try one last time
+    echo "Unable to mount /boot/rpi."
+    mkdir -p /boot/rpi
+    mount -a || exit 1
```
```suggestion
    mount -a || { echo "Failed to mount all partitions. Your kernel is not updated." ; exit 1 ; }
```

You may have to give extra information to the user. (Do not remove the extra semicolon and preserve the spaces.)

Cyanoxygen commented on runtime-kernel/linux-kernel-rpi64-lts/autobuild/postinst.in:
```diff
@@ -0,0 +1,33 @@
+  echo "Updating kernel..."
+  cp -r /usr/lib/rpi64/kernel/* /boot/rpi/
+else
+  echo "Can not find mount point /boot/rpi. Skip copying kernel."
```
For now please treat any issues that unable to mount the boot partition as errors. Exit immediately with a non-zero status (`exit 1`) if any of these errors occured.

Cyanoxygen commented on runtime-kernel/rpi-firmware-boot/autobuild/overrides/usr/lib/rpi64/config/distcfg.txt:
```diff
@@ -0,0 +1,12 @@
+arm_64bit=1
+
+# Kernel image name
+kernel=kernel8.img
```
Unfortunately we have to build two kernels in this package.

More on that later.

Cyanoxygen commented on runtime-kernel/rpi-firmware-boot/spec:
```diff
@@ -0,0 +1,5 @@
+REL=0
+COMMIT="4bb9d889a9d48c7335ebe53c0b8be83285ea54b1"
+SRCS="tbl::https://github.com/raspberrypi/firmware/archive/$COMMIT.tar.gz"
+CHKSUMS="sha256::c2da24fd153b769bd759b11666167a284eaa53cea8f8cc68ea42dc1bf697a80e"
```
Please add a `CHKUPDATE=` section to this package (as the upstream uses tags correctly on this repository).

Cyanoxygen commented on runtime-kernel/rpi-firmware-boot/spec:
```diff
@@ -0,0 +1,5 @@
+VER="1.20241126"
+REL=0
+COMMIT="4bb9d889a9d48c7335ebe53c0b8be83285ea54b1"
+SRCS="tbl::https://github.com/raspberrypi/firmware/archive/$COMMIT.tar.gz"
```
The upstream now uploads proper tarballs in their releases, see https://github.com/raspberrypi/firmware/releases/download/1.20241126/raspi-firmware_1.20241126.orig.tar.xz .

You may want to remove the `COMMIT` and use `${VER}` instead in the urls.

Cyanoxygen commented on runtime-kernel/rpi-firmware-boot/spec:
```diff
@@ -0,0 +1,5 @@
+VER="1.20241126"
```
Please use `UPSTREAM_VER` to record the unmodified version string from upstream; This is for automated version checking.

And accordingly, use `VER="${UPSTREAM_VER}"` to reference the upstream version string in `VER=`.

Cyanoxygen commented on runtime-kernel/linux+kernel+rpi64+lts/autobuild/defines:
```diff
@@ -0,0 +1,14 @@
+
+PKGEPOCH=1
+
+ABRPMAUTOPROVONLY=yes
```
This is no longer used, please remove this line.

Cyanoxygen commented on runtime-kernel/linux+kernel+rpi64+lts/autobuild/defines:
```diff
@@ -0,0 +1,14 @@
+ABSTRIP=no
+ABELFDEP=no
+
+PKGEPOCH=1
```
As there is no package with the same name in the main repository, you can safely remove this line (we are introducing it to the ABBS tree, not reviving it.)

Cyanoxygen commented on runtime-kernel/linux+kernel+rpi64+lts/autobuild/defines:
```diff
@@ -0,0 +1,14 @@
+
+PKGREP="linux+image+new linux+header+new linux+image linux+header linux+image+old linux+header+old"
+PKGPROV="linux+kernel linux+image+new linux+header+new linux+image linux+header linux+image+old linux+header+old"
+
```
As this is a metapackage, you may have to specify `ABTYPE=dummy`.

Cyanoxygen commented on runtime-kernel/linux-kernel-rpi64-lts/autobuild/defines:
```diff
@@ -0,0 +1,16 @@
+PKGBREAK="
+	zfs<=2.2.0
+	nvidia<=530.41.03
+"
```
Please add a new line at the end of the file. Vim, emacs and nano takes care of this, but VS Code doesn't.

If you are using VS Code (or derivatives), Modify the setting `files.insertFinalNewline` in the settings.

Cyanoxygen commented:
Since the Raspberry Pi 5 came out, its kernel has changed significantly. The Linux kernel for Raspberry Pi 5 contains [optimizations specific to it](https://www.raspberrypi.com/documentation/computers/config_txt.html#kernel:~:text=The%20Raspberry%20Pi%205%20firmware%20defaults%20to%20loading%20kernel_2712.img%20because%20this%20image%20contains%20optimisations%20specific%20to%20Raspberry%20Pi%205%20(e.g.%2016K%20page%2Dsize).%20If%20this%20file%20is%20not%20present%2C%20then%20the%20common%2064%2Dbit%20kernel%20(kernel8.img)%20will%20be%20loaded%20instead.), so we have to ship two kernels: one for Pi 4, one for Pi 5.

We can not split the package into two however, as this will break the compatibility with Raspberry Pi 4 (if we only ship the Raspberry Pi 5 kernel - 16K kernel will not boot on Pi 4), or users may experience poor performance (booting 4K kernel on Pi 5, which may suffer from performance issues).

Here I provide you with a plan:

Prepare two configs
--------

The default config name for the Raspberry Pi 4 is `bcm2711_defconfig`. Use this config as the base confg, and set the required options as builtin (`y`):

- `CONFIG_XFS_FS`, `CONFIG_XFS_ONLINE_SCRUB`
- `CONFIG_BTRFS_FS`
- `CONFIG_FW_LOADER`, `CONFIG_FW_LOADER_COMPRESS`, `CONFIG_FW_LOADER_COMPRESS_XZ` and `CONFIG_FW_LOADER_COMPRESS_ZSTD`

And use the local version string `-aosc-rpi64-rpi4`. Install this config to `autobuild/config-rpi4`.

For the config for Raspberry Pi 5, use `bcm2712_defconfig` as the base config, builtin the required options, and use `-aosc-rpi64-rpi5` as the local version string. Install the Pi 5 config to `autobuild/config-rpi5`.

The build script
--------

The build script should be modified accordingly:

1. Copy `autobuild/config-rpi4` to `.config`, run `make` and install the kernel and modules.
2. Run `make mrproper`.
3. Copy `autobuild/config-rpi5` to `.config`, run `make` and install the kernel and modules.
4. Install the device tree blobs and overlay blobs.

Cyanoxygen commented:
Notes
-----

For Raspberry Pi 4, the kernel image name must be `kernel8.img`, and `kernel_2712.img` for the Raspberry Pi 5.

Also you have to remove the `kernel` directive from `distcfg.txt`. The firmware will automatically choose which kernel to boot.

Dustymind committed:
rpi-firmware-boot: update spec and distcfg.txt

Dustymind committed:
linux+kernel+rpi64+lts: clean up defines

Dustymind committed:
linux-kernel-rpi64-lts: edit build scripts

    - Build Raspberry Pi 4 and Raspberry Pi 5 kernels into one package
    - Use upstream config

Dustymind commented:
One problem:
If we run `make mrproper` after Raspberry Pi 4 kernel is built, we will have imcomplete kernel modules for this kernel.
However, the Raspberri Pi 5 kernel modules are not affected.

A temporary workaround will be commited later.

EDIT: Commit was pushed, but further test is required.

<details>
<summary> Details below: </summary>

```
root@main-2bc50594 [ / ] # ls -l /usr/lib/modules/6.6.62-aosc-rpi64-rpi*
/usr/lib/modules/6.6.62-aosc-rpi64-rpi4:
total 180
lrwxrwxrwx  1 root root    45 Dec 11 22:57 build -> /usr/src/linux-headers-6.6.62-aosc-rpi64-rpi4
drwxr-xr-x 11 root root  4096 Dec 12 12:45 kernel
-rw-r--r--  1 root root    45 Dec 12 12:48 modules.alias
-rw-r--r--  1 root root    12 Dec 12 12:48 modules.alias.bin
-rw-r--r--  1 root root 16796 Dec 11 22:57 modules.builtin
-rw-r--r--  1 root root  9028 Dec 12 12:48 modules.builtin.alias.bin
-rw-r--r--  1 root root 18954 Dec 12 12:48 modules.builtin.bin
-rw-r--r--  1 root root 98899 Dec 11 22:57 modules.builtin.modinfo
-rw-r--r--  1 root root     0 Dec 12 12:48 modules.dep
-rw-r--r--  1 root root    12 Dec 12 12:48 modules.dep.bin
-rw-r--r--  1 root root     0 Dec 12 12:48 modules.devname
-rw-r--r--  1 root root    55 Dec 12 12:48 modules.softdep
-rw-r--r--  1 root root    49 Dec 12 12:48 modules.symbols
-rw-r--r--  1 root root    12 Dec 12 12:48 modules.symbols.bin
lrwxrwxrwx  1 root root    45 Dec 11 22:57 source -> /usr/src/linux-headers-6.6.62-aosc-rpi64-rpi4

/usr/lib/modules/6.6.62-aosc-rpi64-rpi5:
total 2760
lrwxrwxrwx  1 root root     45 Dec 11 23:07 build -> /usr/src/linux-headers-6.6.62-aosc-rpi64-rpi5
drwxr-xr-x 11 root root   4096 Dec 12 12:45 kernel
-rw-r--r--  1 root root 654914 Dec 12 12:48 modules.alias
-rw-r--r--  1 root root 681565 Dec 12 12:48 modules.alias.bin
-rw-r--r--  1 root root  16796 Dec 11 23:07 modules.builtin
-rw-r--r--  1 root root   9028 Dec 12 12:48 modules.builtin.alias.bin
-rw-r--r--  1 root root  18954 Dec 12 12:48 modules.builtin.bin
-rw-r--r--  1 root root  98899 Dec 11 23:07 modules.builtin.modinfo
-rw-r--r--  1 root root 231462 Dec 12 12:48 modules.dep
-rw-r--r--  1 root root 313019 Dec 12 12:48 modules.dep.bin
-rw-r--r--  1 root root    356 Dec 12 12:48 modules.devname
-rw-r--r--  1 root root  71677 Dec 11 23:07 modules.order
-rw-r--r--  1 root root    878 Dec 12 12:48 modules.softdep
-rw-r--r--  1 root root 313743 Dec 12 12:48 modules.symbols
-rw-r--r--  1 root root 379263 Dec 12 12:48 modules.symbols.bin
lrwxrwxrwx  1 root root     45 Dec 11 23:07 source -> /usr/src/linux-headers-6.6.62-aosc-rpi64-rpi5
```

</details>

Dustymind committed:
linux-kernel-rpi64-lts: modify build script

    - Fix broken kernel modules build for Raspberry Pi 4
    - typo

Dustymind committed:
rpi-firmware-boot: fix build

Dustymind committed:
rpi-firmware-boot: edit build script

    - treat any mount point errors properly (use 'exit 1' instead of 'exit 0')
    - some adaptation for Raspberry Pi 5

Dustymind committed:
rpi-firmware-boot: fix build

Dustymind committed:
linux-kernel-rpi64-lts: CONFIG_IKCONFIG=y

Cyanoxygen commented on runtime-kernel/linux-kernel-rpi64-lts/autobuild/build:
```diff
@@ -0,0 +1,111 @@
+elif [[ "${CROSS:-$ARCH}" = "ppc64el" ]]; then
+    SRCARCH="powerpc"    
+fi
+
```
You can remove this block - we are not going to build on non-arm64 machines anyway.

Cyanoxygen commented on runtime-kernel/linux-kernel-rpi64-lts/autobuild/build:
```diff
@@ -0,0 +1,111 @@
+
+abinfo "Copying kernel..."
+# make a local copy of compiled kernel
+mkdir -p "$PKGDIR/usr/lib/aosc-os-arm64-boot/linux-kernel-$version$LOCALNAME/"
```
You can remove this line and the line below installing the kernel to this directory. It is not being used, since it was an old standard.

Cyanoxygen commented on runtime-kernel/linux-kernel-rpi64-lts/autobuild/build:
```diff
@@ -0,0 +1,111 @@
+# make a local copy of compiled kernel
+mkdir -p "$PKGDIR/usr/lib/aosc-os-arm64-boot/linux-kernel-$version$LOCALNAME/"
+mkdir -p "$PKGDIR/usr/lib/rpi64/kernel/overlays"
+cp -v arch/arm64/boot/Image "$PKGDIR/usr/lib/aosc-os-arm64-boot/linux-kernel-$version$LOCALNAME/"
```
Ditto.

Cyanoxygen commented on runtime-kernel/linux-kernel-rpi64-lts/autobuild/build:
```diff
@@ -0,0 +1,111 @@
+
+# RPi4 kernel build ends.
+
+# Since `make mrproper` will destroy the installed kernel modules, we need to make a backup then.
+abinfo "Making backup for kernel and kernel modules..."
+tar -cvf rpi4-kernel.tar $PKGDIR
```
@MingcongBai Can we set `PKGDIR` to the parent directory? I tested with that, and it worked well.

Cyanoxygen commented on runtime-kernel/linux-kernel-rpi64-lts/autobuild/build:
```diff
@@ -0,0 +1,111 @@
+
+abinfo "Copying kernel..."
+# make a local copy of compiled kernel
+mkdir -p "$PKGDIR/usr/lib/aosc-os-arm64-boot/linux-kernel-$version$LOCALNAME/"
```
Ditto.

Cyanoxygen commented on runtime-kernel/linux-kernel-rpi64-lts/autobuild/build:
```diff
@@ -0,0 +1,111 @@
+# make a local copy of compiled kernel
+mkdir -p "$PKGDIR/usr/lib/aosc-os-arm64-boot/linux-kernel-$version$LOCALNAME/"
+mkdir -p "$PKGDIR/usr/lib/rpi64/kernel/overlays"
+cp -v arch/arm64/boot/Image "$PKGDIR/usr/lib/aosc-os-arm64-boot/linux-kernel-$version$LOCALNAME/"
```
Ditto.

Cyanoxygen commented on runtime-kernel/linux-kernel-rpi64-lts/autobuild/build:
```diff
@@ -0,0 +1,111 @@
+abinfo "Copying kernel..."
+# make a local copy of compiled kernel
+mkdir -p "$PKGDIR/usr/lib/aosc-os-arm64-boot/linux-kernel-$version$LOCALNAME/"
+mkdir -p "$PKGDIR/usr/lib/rpi64/kernel/overlays"
```
```suggestion
mkdir -p "$PKGDIR/usr/lib/rpi64/kernel/"$version"/overlays"
```

Since multiple kernel versions may be installed at the same time, we should use separate directories to store the kernel. (It is an oversight, sorry.)

Cyanoxygen commented on runtime-kernel/linux-kernel-rpi64-lts/autobuild/build:
```diff
@@ -0,0 +1,111 @@
+mkdir -p "$PKGDIR/usr/lib/aosc-os-arm64-boot/linux-kernel-$version$LOCALNAME/"
+mkdir -p "$PKGDIR/usr/lib/rpi64/kernel/overlays"
+cp -v arch/arm64/boot/Image "$PKGDIR/usr/lib/aosc-os-arm64-boot/linux-kernel-$version$LOCALNAME/"
+cp -v arch/arm64/boot/Image "$PKGDIR/usr/lib/rpi64/kernel/kernel_2712.img"
```
```suggestion
install -Dvm755 "$SRCDIR"/arch/arm64/boot/Image "$PKGDIR/usr/lib/rpi64/kernel/"$version"/kernel_2712.img"
```

Ditto.

Cyanoxygen commented on runtime-kernel/linux-kernel-rpi64-lts/autobuild/build:
```diff
@@ -0,0 +1,111 @@
+cp -v arch/arm64/boot/Image "$PKGDIR/usr/lib/rpi64/kernel/kernel_2712.img"
+
+export ARCH=$OLDARCH
+rm -rf "$PKGDIR"/usr/lib/firmware
```
```suggestion
abinfo "Removing the unnecessary firmware in PKGDIR ..."
rm -rf "$PKGDIR"/usr/lib/firmware
```

A bit of output may help others to know what is going on :-)

Cyanoxygen commented on runtime-kernel/linux-kernel-rpi64-lts/autobuild/build:
```diff
@@ -0,0 +1,111 @@
+abinfo "Copying kernel..."
+# make a local copy of compiled kernel
+mkdir -p "$PKGDIR/usr/lib/aosc-os-arm64-boot/linux-kernel-$version$LOCALNAME/"
+mkdir -p "$PKGDIR/usr/lib/rpi64/kernel/overlays"
```
```suggestion
mkdir -p "$PKGDIR/usr/lib/rpi64/kernel/"$version"/overlays"
```

Since multiple kernel versions may be installed at the same time, we should use separate directories to store the kernel. (It is an oversight, sorry.)

Not using LOCALNAMEs since Pi 4 and Pi 5 are not using the same filename for kernels..

Cyanoxygen commented on runtime-kernel/linux-kernel-rpi64-lts/autobuild/build:
```diff
@@ -0,0 +1,111 @@
+mkdir -p "$PKGDIR/usr/lib/aosc-os-arm64-boot/linux-kernel-$version$LOCALNAME/"
+mkdir -p "$PKGDIR/usr/lib/rpi64/kernel/overlays"
+cp -v arch/arm64/boot/Image "$PKGDIR/usr/lib/aosc-os-arm64-boot/linux-kernel-$version$LOCALNAME/"
+cp -v arch/arm64/boot/Image "$PKGDIR/usr/lib/rpi64/kernel/kernel8.img"
```
```suggestion
install -Dvm755 "$SRCDIR"/arch/arm64/boot/Image "$PKGDIR/usr/lib/rpi64/kernel/"$version"/kernel8.img"
```

Using `install` makes more sense.   
(Also an oversight made by me a few yrs ago) Please use absolute paths.  
And ditto.

Cyanoxygen commented on runtime-kernel/linux-kernel-rpi64-lts/autobuild/build:
```diff
@@ -0,0 +1,111 @@
+# Only copy after both RPi4 and RPi5 kernels and modules were copied.
+
+abinfo "Coping new device tree and overlays to /usr/lib/rpi64/kernel..."
+cp -v arch/arm64/boot/dts/broadcom/*.dtb "$PKGDIR/usr/lib/rpi64/kernel/"
```
```suggestion
cp -v "$SRCDIR"/arch/arm64/boot/dts/broadcom/*.dtb "$PKGDIR"/usr/lib/rpi64/kernel/"$version"/
```

Cyanoxygen commented on runtime-kernel/linux-kernel-rpi64-lts/autobuild/build:
```diff
@@ -0,0 +1,111 @@
+
+abinfo "Coping new device tree and overlays to /usr/lib/rpi64/kernel..."
+cp -v arch/arm64/boot/dts/broadcom/*.dtb "$PKGDIR/usr/lib/rpi64/kernel/"
+cp -rv arch/arm64/boot/dts/overlays/*.dtb* "$PKGDIR/usr/lib/rpi64/kernel/overlays/"
```
```suggestion
cp -rv "$SRCDIR"/arch/arm64/boot/dts/overlays/*.dtb* "$PKGDIR"/usr/lib/rpi64/kernel/"$version"/overlays/
```

(Note how the quotes are being used.)

Cyanoxygen commented on runtime-kernel/linux-kernel-rpi64-lts/autobuild/postinst.in:
```diff
@@ -0,0 +1,43 @@
+      echo "/boot/rpi is created and mounted. Continuing."
+    fi
+    echo "Updating kernel..."
+    cp -r /usr/lib/rpi64/kernel/* /boot/rpi/
```
```suggestion
    cp -r /usr/lib/rpi64/kernel/"$VER"/  /boot/rpi/
```

Cyanoxygen commented on runtime-kernel/linux-kernel-rpi64-lts/autobuild/postinst.in:
```diff
@@ -0,0 +1,43 @@
+  /etc/kernel/postinst.d/apt-auto-removal
+fi
+
+if [ -d "/tree/runtime-kernel" ]; then
```
```suggestion
if ! grep -q /boot/rpi /proc/mounts ; then
```


You can using the following command to detect if `/boot/rpi` is mounted:

```
grep -q /boot/rpi /proc/mounts
```

And you can use `systemd-detect-virt` to detect if we are running in a container (not chroot):
```
systemd-detect-virt -c
```

Cyanoxygen commented on runtime-kernel/rpi-firmware-boot/autobuild/postinst:
```diff
@@ -0,0 +1,58 @@
+#!/bin/bash
+
+if [ -d "/tree/runtime-kernel" ]; then
```
Ditto.
```suggestion
if systemd-detect-virt -q -c ; then
```

Cyanoxygen commented on runtime-kernel/rpi-firmware-boot/autobuild/postinst:
```diff
@@ -0,0 +1,58 @@
+  exit 1
+fi
+
+if [ -e /boot/rpi/kernel8.img ]; then
+  echo "Copying new firmware..."
+  cp -r /usr/lib/rpi64/boot/* /boot/rpi/
+elif [ -e /boot/rpi/kernel_2712.img ]; then
+  echo "Copying new firmware..."
+  cp -r /usr/lib/rpi64/boot/* /boot/rpi/
+else
+  echo "Copying kernel and firmware..."
+  cp -r /usr/lib/rpi64/boot/* /boot/rpi/
+  if [ -e /usr/lib/rpi64/kernel/kernel8.img ]; then
+    cp -r /usr/lib/rpi64/kernel/* /boot/rpi/
+  elif [ -e /usr/lib/rpi64/kernel/kernel_2712.img ]; then
+    cp -r /usr/lib/rpi64/kernel/* /boot/rpi/
+  fi
+fi
```
Since we have to put the kernels in their separate directories and we can not know the version of the kernel installed on the system, so I think this entire logic can be simplified to install the firmware only and unconditionally:

```bash
echo "Installing the boot firmware ..."
cp -r /usr/lib/rpi64/boot/* /boot/rpi/
```

Cyanoxygen commented on runtime-kernel/rpi-firmware-boot/autobuild/postinst:
```diff
@@ -0,0 +1,58 @@
+fi
+
+if [ ! -e /boot/rpi/cmdline.txt ]; then
+  PARTUUID="$(lsblk -x MOUNTPOINT -o MOUNTPOINT,PARTUUID | egrep '^/\s' | awk '{print $2}')"
```
```suggestion
  echo "Getting information of the root partition ..."
  eval "$(findmnt -Pyo PARTUUID /)"
```

Here's the reliable way to find the PARTUUID of the partition which the root filesystem resides on.

Cyanoxygen commented on runtime-kernel/rpi-firmware-boot/spec:
```diff
@@ -0,0 +1,6 @@
+REL=0
+SRCS="tbl::https://github.com/raspberrypi/firmware/releases/download/$VER/raspi-firmware_$VER.orig.tar.xz"
+CHKSUMS="sha256::020dbcdbb30af5942a62fc3eb355449aba45276b67e864dee2522ff53fd936e6"
+CHKUPDATE="anitya::id=21744"
```
Please add a blank line at the end of file.

Cyanoxygen reviewed (changes requested):
I've tested your kernels for both Pi 4 and Pi 5. Good job done! But here's the hard part.

Cyanoxygen commented:
You may want to use the "Files changed" tab to examine the reviews.

Dustymind committed:
rpi-firmware-boot: clean up build script

Dustymind committed:
linux-kernel-rpi64-lts: clean up build scripts

Cyanoxygen commented on runtime-kernel/linux-kernel-rpi64-lts/autobuild/prepare:
```diff
@@ -0,0 +1,8 @@
+ln -sv /usr/bin/ld.bfd "$SRCDIR"/ld
+export PATH="$SRCDIR:$PATH"
+
+sed -e 's|^EXTRAVERSION = -rc.*|EXTRAVERSION =|g' -i Makefile
```
```suggestion
abinfo "Setting PKGDIR to the parent directory ..."
export PKGDIR="$(realpath $SRCDIR/../abdist)"
```


You can remove this `sed` - we are not going to build -rc kernels anyway, and as @MingcongBai stated, you can set `PKGDIR` to the parent of the kernel source tree to avoid compiled modules being deleted by `make mrproper`.

![image](https://github.com/user-attachments/assets/7bf73886-37f3-4aaf-834c-633ce341cd4c)

Cyanoxygen commented on runtime-kernel/linux-kernel-rpi64-lts/autobuild/build:
```diff
@@ -0,0 +1,93 @@
+
+# RPi4 kernel build ends.
+
+# Since `make mrproper` will destroy the installed kernel modules, we need to make a backup then.
+abinfo "Making backup for kernel and kernel modules..."
+tar -cvf rpi4-kernel.tar $PKGDIR
+
```
Since we have set the `PKGDIR` to the parent of the kernel tree, you can safely remove the backup and restore process.

Cyanoxygen commented on runtime-kernel/linux-kernel-rpi64-lts/autobuild/build:
```diff
@@ -0,0 +1,93 @@
+abinfo "Removing all generated files + config + various backup files..."
+make mrproper
+
+# Now, restore the backup.
+abinfo "Restoring the backup..."
+tar -xvf rpi4-kernel.tar --directory=/
```
Ditto.

Dustymind committed:
linux-kernel-rpi64-lts: fix build scripts

    - set PKGDIR to the parent directory
    - remove temporary workaround

Cyanoxygen commented:
Thank you for your contribution! Ciallo～(∠・ω< )⌒☆
